### PR TITLE
Entity parser and converter update

### PR DESCRIPTION
**Pull request by deepu105**
@MathieuAA its done. We need to write more tests and might need more polish here and there.

could you test it out when you have time. the converter produces an array with name and body of entity in jhispter format

### CODE REVIEW

**Comment by MathieuAA** [comment]
Will do

Le 22 juin 2016 10:17:43 GMT+02:00, Deepu K Sasidharan <notifications@github.com> a écrit :
>@MathieuAA its done. We need to write more tests and might need more
>polish here and there.
>
>could you test it out when you have time. the converter produces an
>array with name and body of entity in jhispter format
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/jhipster/jhipster-domain-language/pull/33
>
>-- Commit Summary --
>
>  * add methods to convert entity data to jhipster format
>  * Added lodash as dependency
>  * Added new exceptions
>  * UNSTABLE, partial migration of the DSL parser
>* Merge branch 'master' of
>https://github.com/jhipster/jhipster-domain-language into dslparser
>  * Added angularSuffix and microservice as binary options
>  * Added exception
>  * Finished relationship handling, options are next
>  * Added DTO and pagination options
>  * Added all the options
>  * merge upstream
>  * remove data
>  * remove types
>  * add editorconfig
>  * polish
>  * Merge branch 'master' into dslparser
>  * Merge branch 'dslparser'
>  * polish
>  * polish
>  * remove old parser
>  * add entity convertor
>  * entity convertor unstable
>  * entity convertor test
>  * polish
>  * polish
>  * fix jdl parser
>  * Fixed imports, and made the parser visible
>  * Tested the pre-parsing conditions
>  * Added some error cases
>* Merge branch 'dslparser' of
>https://github.com/jhipster/jhipster-domain-language into dslparser
>  * Removed duplicate
>  * remove duplicate export
>  * fix npm test
>* Merge branch 'dslparser' of
>https://github.com/jhipster/jhipster-domain-language into dslparser
>  * fix convertor tests
>  * add parser tests
>* Merge branch 'master' of
>https://github.com/jhipster/jhipster-domain-language into dslparser
>* Merge branch 'dslparser' of
>https://github.com/jhipster/jhipster-domain-language into dslparser
>  * Removed chalk (#32)
>* Merge branch 'master' of
>https://github.com/jhipster/jhipster-domain-language into dslparser
>* Merge branch 'master' of
>https://github.com/jhipster/jhipster-domain-language into dslparser
>  * change angularSuffixes to angularSuffix fix #31
>  * assign entity options
>  * assign entity fields
>  * assign entity from relations
>  * assign entity from relations
>  * polish
>  * polish and fixes
>  * polish
>  * rename tests
>
>-- File Changes --
>
>    A .editorconfig (11)
>    A lib/convertor/jdl_convertor.js (368)
>    M lib/core/jdl_relationship.js (53)
>    M lib/core/jdl_relationships.js (18)
>    M lib/core/jhipster/relationship_types.js (4)
>    M lib/dsl/grammar.txt (83)
>    A lib/dsl/jdl_parser.js (225)
>    M lib/dsl/pegjs_parser.js (277)
>    M lib/exceptions/exception_factory.js (5)
>    M lib/reader/jdl_reader.js (42)
>    A lib/utils/format_utils.js (51)
>    M lib/utils/object_utils.js (43)
>    M lib/utils/objects/set.js (4)
>    M module/index.js (6)
>    M package.json (3)
>    A test/spec/convertor/jdl_convertor_test.js (41)
>    A test/spec/dsl/jdl_parser_test.js (63)
>    R test/spec/reader/jdl_reader_test.js (28)
>    A test/test_files/complex_jdl.jdl (116)
>
>-- Patch Links --
>
>https://github.com/jhipster/jhipster-domain-language/pull/33.patch
>https://github.com/jhipster/jhipster-domain-language/pull/33.diff
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/jhipster/jhipster-domain-language/pull/33

-- 
Mathieu ABOU-AICHI

**Comment by MathieuAA** [comment]
There's still some minor work to be done (exceptions, not using the `#forEach` method, etc.), but otherwise it's good! Tests are to be done too. I'll see if I can make some progress tonight.

**Comment by MathieuAA** [comment]
The convertor needs a complete rehaul... but first, the tests for the DSL parser have to be done.

File: lib/core/jdl_relationships.js [hunk 1]
```diff
@@ -22,26 +23,23 @@ class JDLRelationships {
       throw new buildException(exceptions.InvalidObject, 'A valid relationship must be passed.');
     }
     this.relationships[relationship.type][relationship.getId()] = relationship;
+    this.size++;
   }
 
-  getAll() {
+  toArray() {
```
Comment (deepu105) [inline lib/core/jdl_relationships.js]: Ok this is better than getAll +1

**Comment by deepu105** [comment]
@MathieuAA actually is there a need for the intermediate format that you create as output from the `jdl_parser`? is it required for JHUML? If not we can actually simplify this to parse the pegJS output directly to below format which can be used by the `import-jdl` sub gen to create entities. If thats the case the core objects model might not be required as well. But I would assume you saw some usecase for that format?
Also is there any specific reason for changing `forEach` to `for in` loops?

```
[
  {
    name: 'foo`,
    body: fooJsonObject
  },
  {
    name: 'bar`,
    body: barJsonObject
  }
]
```
Did i post this comment already? I remember doing it but didnt find it here :wink:

**Comment by MathieuAA** [comment]
Yes, it will be used by JUML (or anyone else for that matter), and another project I hope to work on very soon. The whole point of whether one wants to deal with the JDL objects rather than what PegJS returns was raised (by myself) a couple of months back. If one doesn't want to deal with the JDL objects, you can simply use the first link in the chain (PegJs -> DSLparser -> Convertor) in the subgen. That choice is up to you! Use whatever you want =)

As for the forEach and for in, the for in version runs faster and is lighter in memory (plus, the GC handles it better because of the let keyword introduced).

**Comment by deepu105** [comment]
for in is better than for each but the use case here is `Object.keys + forEach` vs `hasOwnProperty check + for in` which might perform worse then the other combination check this http://stackoverflow.com/questions/25052758/for-in-vs-object-key-foreach-without-inherited-properties

If you are keen on performance use `Object.keys + for loop` as the thread suggests.

**Comment by MathieuAA** [comment]
Thanks, I'll check it ASAP. By the way, thanks @jdubois for changing the project's name =) and @deepu105 for changing the package.json file

**Comment by MathieuAA** [comment]
We'll have to do some benchmarking, but I like the 
```
for (let i = 0, keys = Object.keys(object); i < keys.length; i++) {
    // ...
}
```
version.

**Comment by deepu105** [comment]
@MathieuAA are you ok with below?

1. move jdl_parser to parser folder
2. move jdl_convertor to parser folder and rename to entity_parser 
3. parse the pegJS output directly to jhipster entity without a the intermediate

or do you prefer the current impl? im fine either way as there is no noticeable performance impact

**Comment by MathieuAA** [comment]
I'm ok with the first two, about the third one: do you mean that the convertor will use what's returned by the PegJS parser? If yes, I'm ok!

**Comment by deepu105** [comment]
On second thought I guess it might be better to have the `converter(entity_parser)` convert from `jdl_object` output by `jdl_parser` so that tomorrow if you want to you convert UML to JDL modal and then the converter(entity_parser) can convert and pass it to `import-jdl` sub generator to create entities. It would save a step of creating a JDL document and then parsing it again. what do you think @MathieuAA ?

**Comment by MathieuAA** [comment]
Sorry, I couldn't answer sooner... Flat tyres are the worst.
Whatever works for me. If choose this solution, we would not expose ourselves to changes to the grammar. Let's use the DSL Parser then =)

**Comment by deepu105** [comment]
@MathieuAA Ok then its settled lets keep as it is. Merge this when you think its ready. let me know if you want me to change anything or polish anything. We need to do a release soon and as soon as its released I'll update the sub generator to use this and we can close the long standing issues.

**Comment by MathieuAA** [comment]
Well, it depends how soon you want it to happen. Tests have to be done to be sure everything's ok, coverage should be high too, Moreover, some issues have to be closed (#26, #25, #21, #19).

**Comment by deepu105** [comment]
#26 and #19 needs to be closed and agree we need to test it well. I guess
once some more tests are added for the jdl_parser and entity_parser we
should have high coverage.

#25 and #21 can wait as first one is a feature request and second one needs
to be reconsidered as there is a bigger issue with that in jhipster itself
see https://github.com/jhipster/generator-jhipster/issues/3723

I'll like to have this released sooner rather than later.

**Comment by MathieuAA** [comment]
#21 and the issue in JHipster are not related, are they?

**Comment by deepu105** [comment]
Not related but we need to see if we are supporting this option or not. IMO
we should remove this option unless its fixed in JH as it will cause more
harm then good


Thanks & Regards,
Deepu

File: lib/parser/entity_parser.js [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+ * @returns {Array} the changed entities.
+ */
+function filterOutUnchangedEntities(entityArray) {
+  return entityArray.filter(function (entity) {
+    var currEntity = onDiskEntities[entity.name];
+    var newEntity = entity.body
+    if (!currEntity) {
+      return true;
+    }
+    return !areEntitiesEqual(currEntity, newEntity);
```
Comment (deepu105) [inline lib/parser/entity_parser.js]: @MathieuAA  why did you remive this method from the util file?  Its not unused

**Comment by MathieuAA** [comment]
Oh, I might I have made a mistake then! As it wasn't tested at all, I decided to remove it.

**Comment by deepu105** [comment]
Ya need to add test

Thanks & regards,
Deepu

**Comment by MathieuAA** [comment]
Closing this as all the work is done on a separate branch... no need for this PR anymore.

**Pull request merged**